core/includes/common.inc

```diff
@@ -7112,101 +7112,72 @@ function element_info_property($type, $property_name, $default = NULL) {
+  foreach ($keys as $key => $key_sort) {
+    if (is_string($key_sort) && is_int($key)) {
+      $new_keys[$key_sort] = SORT_NUMERIC;
     }
     else {
-      if ($sort === SORT_NUMERIC || $sort === SORT_STRING) {
-        $new_keys[$index] = $sort;
+      if ($key_sort === SORT_NUMERIC || $key_sort === SORT_STRING) {
+        $new_keys[$key] = $key_sort;
       }
       else {
         // Fallback behavior for unexpected values. Untranslated to allow this
         // function to be used anywhere within Backdrop.
-        $new_keys[$index] = SORT_NUMERIC;
+        $new_keys[$key] = SORT_NUMERIC;
         trigger_error('backdrop_sort() expects the second parameter to be an array keyed by strings and each value of the array to be either SORT_NUMERIC or SORT_STRING.', E_USER_WARNING);
       }
     }
   }
   $keys = $new_keys;
 
-  // If sorting on a single value, optimize the sorting to sort only on that
-  // specific key without the overhead of recursive handling.
-  if (count($keys) === 1) {
-    $key = key($keys);
-    $key_sort = reset($keys);
-
-    if ($key_sort === SORT_STRING) {
-      uasort($array, function($a, $b) use ($key, $dir) {
-        if (!is_array($a) || !is_array($b)) {
-          return 0;
-        }
-        if (!isset($a[$key])) {
-          $a[$key] = '';
-        }
-        if (!isset($b[$key])) {
-          $b[$key] = '';
-        }
-        $cmp = strnatcasecmp($a[$key], $b[$key]);
-        return ($dir == SORT_DESC) ? -$cmp : $cmp;
-      });
-    }
-    else {
-      uasort($array, function($a, $b) use ($key, $dir) {
-        if (!is_array($a) || !is_array($b)) {
-          return 0;
-        }
-        if (!isset($a[$key])) {
-          $a[$key] = 0;
-        }
-        if (!isset($b[$key])) {
-          $b[$key] = 0;
-        }
-        if ($a[$key] != $b[$key]) {
-          $cmp = $a[$key] < $b[$key] ? -1 : 1;
-          return ($dir == SORT_DESC) ? -$cmp : $cmp;
-        }
-        // Equal values, return 0.
-        return 0;
-      });
-    }
-  }
-  // If doing a multiple-key comparison, use a recursive callback.
-  else {
-    $keys_map = array_keys($keys);
-    $recursive_callback = function ($a, $b, $key_index = 0) use ($keys, $keys_map, $dir, &$recursive_callback) {
-      $key = $keys_map[$key_index];
-      if (!is_array($a) || !is_array($b)) {
-        return 0;
-      }
-      if (!isset($a[$key])) {
-        $a[$key] = 0;
-      }
-      if (!isset($b[$key])) {
-        $b[$key] = 0;
-      }
-      if ($a[$key] == $b[$key]) {
-        // Sort on the next key if one exists.
-        if (isset($keys_map[$key_index + 1])) {
-          return $recursive_callback($a, $b, $key_index + 1);
-        }
+  // Create an array to use for sorting by each key.
+  $sort = array();
+  foreach ($array as $k => $v) {
+    $v = (array) $v;
+    foreach ($keys as $key => $key_sort) {
+      if (isset($v[$key])) {
+        $sort[$key][$k] = ($key_sort === SORT_STRING) ? backdrop_strtolower($v[$key]) : $v[$key];
```

Comment:
@BWPanda it looks like if we're certain that the strings are going to be ASCII only then we can use `strtolower()` otherwise `backdrop_strtolower()` is the best option. I guess we can't predict that every use case will be ASCII only.

Reply:
> I guess we can't predict that every use case will be ASCII only.

Yeah, and with the possibility for emoji in the database, eventually they might make it into core... So since this doesn't seem like a regression performance-wise, I'd say leave it as `backdrop_strtolower()`.